### ISSUE TITLE
Implement abstract clear packet worker

### ISSUE THREAD

**Issue by soareschen** [informalsystems/hermes#3131]
## Summary

The new relayer supports event-based reactive relaying through the `CanRelayEvent` trait and event-based auto relayers such as `ParallelEventSubscriptionRelayer`. In this mode, the packet relaying only works if the relayer receives the packet event from the underlying event stream provided by `HasEventSubscription`. As a result, the relayer may miss packet events that are emitted *before* the relayer is started or due to reasons such as network disconnection from the event source.

To support relaying of packets that the event-based auto relayers missed, clear packet functionality similar to the current relayer needs to be implemented.

## Packet clearing in current relayer

The core logic for the clear packet functionality of the current relayer resides in several functions in the `ibc_relayer::link::relay_path` module. In `schedule_recv_packet_and_timeout_msgs`, the function calls `query_unreceived_packets` to query from the source chain for the list of packets that are not yet received by the destination chain. In `schedule_packet_ack_msgs`, the function calls `query_unreceived_acknowledgements` to query from the destination chain for the list of packets whose ack has not yet been received by the source chain. The functions put the queried packets into the operational data queue so that they are bundled into batches of transactions that the relayer would submit to the chain. These functions are called periodically by the `relay_pending_packets` to relay any packet that it has missed from the event-based relaying.

## Implementation strategy

For the new relayer, clear packet worker can be implemented in simpler ways. Since the messages are batched separately by the batch message worker, multiple clear packet workers can be run in parallel with the event-based auto relayers. The implementation strategy is as follows:

### Querier Components

Querier component traits such as `CanQueryUnreceivedPackets` and `CanQueryUnreceivedAcks` would be defined. The rough implementation outline is as follows:

```rust
#[async_trait]
pub trait CanQueryUnreceivedPackets<Counterparty>:
    HasIbcPacketTypes<Counterparty>
where
    Counterparty: HasIbcPacketTypes<
        Self,
        IncomingPacket = Self::OutgoingPacket,
        OutgoingPacket = Self::IncomingPacket,
    >,
{
    async fn query_unreceived_packets(
        &self,
        channel_id: &Self::ChannelId,
        port_id: &Self::PortId,
        sequences: &[Counterparty::Sequence],
        /* ... */
    ) -> Result<Vec<Self::IncomingPacket>, Self::Error>;
}

#[async_trait]
pub trait CanQueryUnreceivedAcks<Counterparty>:
    HasIbcPacketTypes<Counterparty>
where
    Counterparty: HasIbcPacketTypes<
        Self,
        IncomingPacket = Self::OutgoingPacket,
        OutgoingPacket = Self::IncomingPacket,
    >,
{
    async fn query_unreceived_acks(
        &self,
        channel_id: &Self::ChannelId,
        port_id: &Self::PortId,
        sequences: &[Self::Sequence],
        /* ... */
    ) -> Result<Vec<Self::OutgoingPacket>, Self::Error>;
}

```

Note that the key difference between `CanQueryUnreceivedPackets` and `CanQueryUnreceivedAcks` is that the `Sequence` types are opposite, and that one result in `OutgoingPacket`, while the other results in `IncomingPacket`.

According to the current implementation, it may also be necessary to first query for packet commitments from the counterparty chain. Additional traits such as follows may be needed:

```rust
#[async_trait]
pub trait CanQueryPacketCommitments<Counterparty>:
    HasIbcPacketTypes<Counterparty>
where
    Counterparty: HasIbcPacketTypes<
        Self,
        IncomingPacket = Self::OutgoingPacket,
        OutgoingPacket = Self::IncomingPacket,
    >,
{
    async fn query_packet_commitments(
        &self,
        channel_id: &Self::ChannelId,
        port_id: &Self::PortId,
        /* ... */
    ) -> Result<Vec<Self::Sequence>, Self::Error>;
}

#[async_trait]
pub trait CanQueryPacketAcks<Counterparty>:
    HasIbcPacketTypes<Counterparty>
where
    Counterparty: HasIbcPacketTypes<
        Self,
        IncomingPacket = Self::OutgoingPacket,
        OutgoingPacket = Self::IncomingPacket,
    >,
{
    async fn query_packet_acknowledgements(
        &self,
        channel_id: &Self::ChannelId,
        port_id: &Self::PortId,
        sequences: &[Counterparty::Sequence],
        /* ... */
    ) -> Result<Vec<Counterparty::Sequence>, Self::Error>;
}

```

Note that the above are just draft implementation. Due to things getting confusing pretty easily, the exact types such as whether a parameter should be `Self::Sequence` or `Counterparty::Sequence` may differ. Additional parameters may also be needed, such as for querying of packets sent before a given height.

### Active auto relayers

The event-based auto relayers can be considered *passive* relayers, while the packet clearing relayers can be considered *active* relayers. The active relayers would be implemented as multiple `AutoRelayer` components:

```rust
pub struct RelayUnreceivedPackets;

#[async_trait]
impl<Relay> AutoRelayer<Relay> for RelayUnreceivedPackets
where
    Relay: CanRelayPacket,
    Relay::DstChain: CanQueryUnreceivedPackets<Relay::SrcChain>,
{
    async fn auto_relay(relay: &Relay) -> Result<(), Relay::Error> {
        // ...

        let packets = relay.destination_chain().query_unreceived_packets(/* ... */)
            .await
            .map_err(Relay::dst_chain_error)?;

        stream::iter(packets)
            .for_each_concurrent(None, |packet| async {
                relay.relay_packet(&packet).await;
            })
            .await
        ;
        Ok(())
    }
}

```

### Interval auto relayers

The auto relayers above such as `RelayUnreceivedPackets` would only automatically perform the clearing of unreceived packets once. To make it perform the same task periodically at a fixed interval, it would be additionally wrapped in an *interval relayer* that periodically calls the inner auto relayer:

```rust
pub trait HasAutoRelayInterval {
    fn auto_relay_interval(&self) -> Duration;
}

pub struct IntervalRelayer<InRelayer>(PhantomData<InRelayer>);

#[async_trait]
impl<Relay, InRelayer> AutoRelayer<Relay> for IntervalRelayer<InRelayer>
where
    Relay: HasRelayTypes + HasRuntime + HasAutoRelayInterval,
    InRelayer: AutoRelayer<Relay>,
    Relay::Runtime: CanSleep,
{
    async fn auto_relay(relay: &Relay) -> Result<(), Relay::Error> {
        let sleep_interval = relay.auto_relay_interval();
        loop {
            relay.runtime().sleep(sleep_interval).await;
            InRelayer::auto_relay(relay).await?;
        }
    }
}

```

This would allow separation of the periodic execution of the clear packet functionality and not mix it with the core logic.

### Modification of existing traits

Some modifications may be needed to help facilitate the clear packet implementation:

- The `auto_relay` trait method should return a `Result<(), Self::Error>`. Errors returned from existing calls to `auto_relay()` can be ignored for now and logged later on.
- The `HasRelayTypes` trait should also provide accessor methods for the channel and port IDs, such as `source_channel_id()`. If feasible, the accessor methods together with the current methods `source_client_id` and `destination_client_id` should be moved to a separate trait called `HasRelayParams`.

## Concrete Implementation

The implementation details above only cover the abstract implementation of the packet relayers. The concrete implementation would involve adding methods to the all-in-one traits, as well as the concrete contexts such as `CosmosChain`.

To better track the work needed, the work for concrete implementation will be tracked in a separate issue. This issue can be closed once the abstract components have been implemented.
